packages/zig2/game.ts

```diff
@@ -0,0 +1,325 @@
+
+    const playcard = playCard(state, data.cards);
+    updateState((draftState) => {
+        const { gameState } = draftState.rooms[roomId];
+        gameState.playedCard = playcard;
+        gameState.players[gameState.currentPlayer].cardSet =
+            state.players[state.currentPlayer].cardSet;
+        gameState.prevPlayer = gameState.currentPlayer;
+        setNextPlayer(draftState, roomId);
+    });
+    const newState = context.StateManager.rooms[roomId].gameState;
+    broadcastBig2State(context, 'GameState', newState);
+    if (checkWin(state.players[state.prevPlayer].cardSet)) {
+        process.stdout.write(state.players[state.prevPlayer].client + ' won the game');
+    }
+};
+
+function isValidPlay(state: State, playIndies: number[]) {
+    const playerCardSet = state.players[state.currentPlayer].cardSet;
+    for (const i of playIndies) {
+        if (i > playerCardSet.length - 1) {
+            process.stdout.write('!! --- player does not have the card --- !!');
+            return false;
+        }
+    }
+    const candidates = [];
+    playIndies = [].slice.call(playIndies).sort((a: number, b: number) => {
+        return b - a;
+    });
+    for (const i of playIndies) {
+        candidates.push(playerCardSet[i]);
+    }
+    const currentPattern = getPattern(candidates);
+    if (currentPattern.type === Pattern.INVALID) {
+        process.stdout.write('!! --- not a valid pattern --- !!');
+        return false;
+    }
+    if (state.currentPlayer === state.prevPlayer) {
+        return true;
+    }
+    const prevPattern = getPattern(state.playedCard);
+    if (
+        currentPattern.type !== prevPattern.type ||
+        currentPattern.priority < prevPattern.priority
+    ) {
+        return false;
+    }
+    return true;
+}
+
+function getPriority(cards: Card[]) {
+    let priority = -1;
+    for (const card of cards) {
+        if (card.priority > priority) {
+            priority = card.priority;
+        }
+    }
+    return priority;
+}
+
+function getPattern(cards: Card[]) {
+    const hierarchy = getPriority(cards);
+    if (cards.length === 1) {
+        return { type: Pattern.SINGLE, priority: hierarchy };
+    } else if (cards.length === 2) {
+        if (cards[0].rank === cards[1].rank) {
+            return {
+                type: Pattern.SINGLE,
+                priority: hierarchy,
+            };
+        }
+        return { type: Pattern.INVALID, priority: -1 };
+    } else if (cards.length === 3) {
+        if (cards[0].rank === cards[1].rank && cards[0].rank === cards[2].rank) {
+            return { type: Pattern.THREE, priority: hierarchy };
+        }
+        return { type: Pattern.INVALID, priority: -1 };
+    } else if (cards.length === 5) {
+        if (isStraightFlush(cards)) {
+            return { type: Pattern.FIVE, priority: getPriorityForStraight(cards) + 500 };
+        } else if (isFourOfAKind(cards)) {
+            return { type: Pattern.FIVE, priority: cards[2].priority + 400 };
```

Comment:
I applied the big2 rule from the wikipedia. The rule is a little different than our cognition. For single, pair,  and three of a kind, the pattern must be match with previous played cards. Rest patterns are in order of straight < flush < fullhouse < 4+1 < straightflush.

The reason why I hard code the priority (100 - 500) is because this priority system is only for BigTow (is not reusable). The ranking system vary widely from other playing ways(斗地主 保皇 ~~~).

Reply:
We can't use sum of the priorities as final priority. For example, 
full house(Three a kind  + pair) the priority is depended on Three a kind :  
let's say Ace=14, 2=15, others are same as rank
In sum up system
44422 =  42,       55566 = 27.
44422 should win. But 55566 is actually having higher priority.

Reply:
Yes, the hand is sorted since dialed. There are some benefits for sorted card sets. client doesn't need to do the sorting. It's easier to determine the pattern of a play with sorted cards. The most importantly, we only need to sort it once at the beginning.